dev/com.ibm.ws.microprofile.reactive.messaging.kafka/src/com/ibm/ws/microprofile/reactive/messaging/kafka/CommittingPartitionTracker.java

```diff
@@ -0,0 +1,304 @@
+            committedOffset = newCommitOffset;
+        }
+    }
+
+    /**
+     * Commit the offset up to the offset of the given work
+     *
+     * @param work
+     * @return completion stage which completes with the result of the commit, or completes exceptionally if the commit failed
+     */
+    private CompletionStage<Void> commitUpTo(CompletedWork work) {
+        if (isClosed()) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(this, tc, "Rejecting commit attempt because partition is closed", this);
+            }
+
+            CompletableFuture<Void> result = new CompletableFuture<>();
+            result.completeExceptionally(new Exception("Partition is closed"));
+            return result;
+        }
+
+        // Note work.offset + 1
+        // Committed offset is the _next_ message we expect to receive
+        OffsetAndMetadata offsetAndMetadata = factory.newOffsetAndMetadata(work.offset + 1, work.leaderEpoch, null);
+        return kafkaInput.commitOffsets(topicPartition, offsetAndMetadata);
+    }
+
+    /**
+     * Complete the CompletionStage associated with work which has been committed
+     * <p>
+     * This is called as callback after the partition offset has been committed asynchronously
+     *
+     * @param originalOffset the committed offset before this commit
+     * @param committedOffset the new committed offset
+     * @param exception the exception which caused the commit to fail, or {@code null} if it was successful
+     */
+    private void processCommittedWork(long originalOffset, long committedOffset, Throwable exception) {
+
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            if (exception == null) {
+                Tr.debug(this, tc, "Commit from " + originalOffset + " to " + committedOffset + " completed successfully", this);
+            } else {
+                Tr.debug(this, tc, "Commit from " + originalOffset + " to " + committedOffset + " failed", this, exception);
+            }
+        }
+
+        // Note: Pull out the list of completed work inside the synchronized block
+        //       but complete the CompletionStage outside the synchronized block
+        List<CompletedWork> committedWork = new ArrayList<>();
+        synchronized (completedWork) {
+            for (Iterator<CompletedWork> i = completedWork.iterator(); i.hasNext();) {
+                CompletedWork completedWork = i.next();
+                if (completedWork.offset < originalOffset) {
+                    continue;
+                }
+                if (completedWork.offset >= committedOffset) {
+                    break;
+                }
+                committedWork.add(completedWork);
+                i.remove();
+            }
+        }
+
+        if (exception == null) {
+            for (CompletedWork completedWork : committedWork) {
+                completedWork.completion.complete(null);
```

Comment:
Maybe use a different local variable name? To avoid confusion with the instance TreeSet variable above.